include/mqtt/broker/session_state.hpp

```diff
@@ -171,6 +171,49 @@ struct session_state {
         return tim_session_expiry_;
     }
 
+    void publish(
+        as::io_context& ioc,
+        buffer pub_topic,
+        buffer contents,
+        publish_options pubopts,
+        v5::properties props) {
+
+        BOOST_ASSERT(online());
+
+        if (offline_messages_.empty()) {
+            try {
+                // TODO: Probably this should be switched to async_publish?
+
+                //       Given the async_client / sync_client seperation
+                //       and the way they have different function names,
+                //       it wouldn't be possible for broker.hpp to be
+                //       used with some hypothetical "async_server" in the future.
+                con_->publish(
+                    pub_topic,
+                    contents,
+                    pubopts,
+                    props
+                );
+                return;
+            }
+            catch (packet_id_exhausted_error const& e) {
+                MQTT_LOG("mqtt_broker", warning)
```

Comment:
Here again, the result of the publish is handled by an exception. This will probably be called less often, but I don't think it is exceptional that the packet_id can run out. Also a warning is generated, but the broker is able to handle this behaviour now. So nothing to warn about. 

Ideally you want to generate the packet_id here by the broker, and call publish with the pre-generated packet id. This should be possible with optional acquire_unique_packet_id

Reply:
I think packet_id_exhausted is exceptional case. I understand it actually happens. Even if it handles well in the broker, finally run out memory and got bad_alloc. And in the broker exception happens just one time across threshold, then the message is go to offline_messages.
If packet_id_exhausted frequently happens, the network capacity design is broken.

Reply:
But as you mentioned, optional version of packet_id acquisition is work considering. Please wait a moment.

Reply:
done :)

Reply:
I agree that network design is broken. If client becomes unresponsive or disconnected, message may accumulate. Or if broker is very slow, also messages accumulate. Ideally the connection should timeout soon. Otherwise the broker may accumulate a lot of memory. But this is use case optimization i think. User should select qos and connection keepalive timeouts to prevent this. Or make option to set maximum of offline messages per session?

Reply:
> I agree that network design is broken. If client becomes unresponsive or disconnected, message may accumulate. Or if broker is very slow, also messages accumulate. Ideally the connection should timeout soon. Otherwise the broker may accumulate a lot of memory. But this is use case optimization i think. User should select qos and connection keepalive timeouts to prevent this. Or make option to set maximum of offline messages per session?

I'm not sure I answer your question but,
- User can use Message Expiry Interval
- User can set Receive Maximum  https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901049, and broker also set it https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901083. It is bidirectional. (Not implemented yet)
- User can use some kind of application level flow control

I think that there are many existing options.

So far, I don't have a plan to implement broker option to limit offline message (size count). But in the future, I will add it if it is really needed.